### ISSUE TITLE
Behaviour of .env in regards to buildDepends changed in 20.03

### ISSUE THREAD

**Issue by sevanspowell** [NixOS/nixpkgs#84070]
**Describe the bug**
Haskell derivations: behaviour of `.env` in regards to `buildDepends` changed in 20.03. Projects relying on the old behaviour will break.

A number of Haskell projects I work on use `pkgs.haskell.lib.addBuildDepends` to make test dependencies and build tools available in the environment when invoking `nix-shell`.

PR https://github.com/NixOS/nixpkgs/pull/76349 changed the implementation of the `.env` attribute which subtly changes the effect `addBuildDepends` has. Using `addBuildTools` instead is a workaround for this (or possibly the function I should've been using all along).

I believe I was using `addBuildDepends` incorrectly but it is difficult to be sure, because the expected usage of `addBuildDepends` and `addBuildTools` is not documented (AFAIK).

Using

```
  drv = pkgs:
    pkgs.haskell.lib.addBuildDepends
      (pkgs.haskellPackages.callCabal2nix "minimal-env-issue" ./. {})
      (with pkgs.haskellPackages; [ ghcid ] ++ (with pkgs; [ minio ]));
```

In 19.09, (drv.env.buildInputs ++ drv.env.nativeBuildInputs) produces:
```
[ "/nix/store/83gw88ngcwyq0y06nmcs12svd6ymk5mp-minio-2019-02-26T19-51-46Z"
, null
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/9sl95r0rrqs8mqs2jd22ydwyi8zs8imb-remove-references-to"
, "/nix/store/rj07rp9iwzn1f69ap6c1nsgghz3w16xy-ghc-8.6.5-with-packages"
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/9sl95r0rrqs8mqs2jd22ydwyi8zs8imb-remove-references-to"
]
```

In 20.03, (drv.env.buildInputs ++ drv.env.nativeBuildInputs) produces:
```
["/nix/store/2j8yfxhhyhs6am6ras9q18rnj1hjqngw-ghc-8.6.5-with-packages"]
```

`minio` is present in 19.09 but not 20.03. Interestingly, `ghcid` is not present in either.

If we instead use `addBuildTools`:

```
  drv = pkgs:
    pkgs.haskell.lib.addBuildTools
      (pkgs.haskellPackages.callCabal2nix "minimal-env-issue" ./. {})
      (with pkgs.haskellPackages; [ ghcid ] ++ (with pkgs; [ minio ]));
```

we get the following results in 19.09:

```
[ null
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/9sl95r0rrqs8mqs2jd22ydwyi8zs8imb-remove-references-to"
, "/nix/store/83gw88ngcwyq0y06nmcs12svd6ymk5mp-minio-2019-02-26T19-51-46Z"
, "/nix/store/rj07rp9iwzn1f69ap6c1nsgghz3w16xy-ghc-8.6.5-with-packages"
, "/nix/store/q4q62h0ab7ag0k161cgqknbp7s9qxfhd-ghc-8.6.5"
, "/nix/store/9sl95r0rrqs8mqs2jd22ydwyi8zs8imb-remove-references-to"
, "/nix/store/mi7hxdc3jllkvn3svvqvcllq72fl0g57-ghcid-0.7.5"
, "/nix/store/83gw88ngcwyq0y06nmcs12svd6ymk5mp-minio-2019-02-26T19-51-46Z"
]
```

and in 20.03:

```
[ "/nix/store/bp9ij10sl87lrsyxmixqd6fyy30l5gxf-ghc-8.6.5"
, "/nix/store/nd4licplxl9p5v1m7y5dpri2f0cj99j5-ghcid-0.7.7"
, "/nix/store/fkjdvl6k71d6xdybvzpd4qlnain01l3w-minio-2019-10-12T01-39-57Z"
]
```

`minio` and `ghcid` are present in both.

So, using `addBuildTools` in `20.03` and `19.03` does exactly what I want (with the `20.03` version being a bit cleaner).

The "fix" for this issue might just be to document the intended usage of: `buildDepends`, `setupHaskellDepends`, `libraryHaskellDepends`, `executableHaskellDepends`, `buildTools`, `libraryToolDepends`, `executableToolDepends`, `testToolDepends` and `benchmarkToolDepends` a bit more clearly (if not already - sorry, I couldn't find any).

**To Reproduce**
1. Download [this gist](https://gist.github.com/sevanspowell/baa3f2cbb973adf21904d2eb5a0578fa) ([direct download](https://gist.github.com/sevanspowell/baa3f2cbb973adf21904d2eb5a0578fa/archive/2975309097c24a3bcdde4cb94f30300fe216c9cd.zip)).
2. unzip it and cd into the directory.
3. Execute the following commands in a `nix-shell`:
  ```
  # To determine if the async haskell package has correctly been added to the environment
  ghc-pkg list | grep async
  # To determine if a haskellPackage added with `addBuildDepends/Tools` is present
  ghcid --version
  # To determine if a regular package added with `addBuildDepends/Tools` is present
  minio version
  ```
  where the `nix-shell` invocation is one of:
  ```
  # 19.09, addBuildDepends
  nix-shell --pure -A drv-19-09.env
  # 20.03, addBuildDepends
  nix-shell --pure -A drv-20-03.env
  # 19.09, addBuildTools
  nix-shell --pure -A drvFix-19-09.env
  # 20.03, addBuildTools
  nix-shell --pure -A drvFix-20.03.env
  ```
3. A number of convenience make targets are available too.

**Expected behavior**
I expected `addBuildDepends` to make `minio` and `ghcid` available in my `.env` shell.

**Comment by cdepillabout** [comment]
cc @Ericson2314 as the author of #76349, @jmininger (who I think wrote the original version), and @matthewbauer (who did an extensive review)

--------------------------------------------

We should probably have documentation for this, or at the very least a note in the release notes for 20.03.

**Comment by codygman** [comment]
Side question: how did you print the result of `addBuildDepends` in the context of your actual package?

**Comment by sevanspowell** [comment]
> Side question: how did you print the result of `addBuildDepends` in the context of your actual package?

So I created a small nix file something like:

file.nix
```
let
  pkgs = import (fetchGit (import ./pkgs.nix)) { };

  drv = pkgs:
    pkgs.haskell.lib.addBuildDepends
      (pkgs.haskellPackages.callCabal2nix "minimal-env-issue" ./. {})
      (with pkgs.haskellPackages; [ ghcid ] ++ (with pkgs; [ minio ]));

  drvFix = pkgs:
    pkgs.haskell.lib.addBuildTools
      (pkgs.haskellPackages.callCabal2nix "minimal-env-issue" ./. {})
      (with pkgs.haskellPackages; [ ghcid ] ++ (with pkgs; [ minio ]));
in
{
  inherit drv drvFix;
}
```

Then:
```
nix repl file.nix
> (drv.env.buildInputs ++ drv.env.nativeBuildInputs)
```

**Comment by Ericson2314** [comment]
I think this looks correct to me. Nevermind the build-time vs run-time distinctions I am usually harping about. The most important thing to understand for this is just like the regular generic builder always did, the `env` now respects the different type of inputs parameters stating Haskell vs non-Haskell, and uses that instead of the predicate it used to sift between the combination of everything before.

tl;dr `buildDepends` and `<comp-flavor>Depends` items always *don't* become `buildInputs` in env, and instead get passed on to the `ghcWithPackages`, while `extraLibraries` and `<comp-flavor>SystemDepends` do become `buildInputs`. On the build-time deps side, including the tools, whether or not something was written in Haskell doesn't matter, so they all become `nativeBuildInputs`.
